channeldb/invoices.go

```diff
@@ -1957,39 +1984,72 @@ func cancelSingleHtlc(resolveTime time.Time, htlc *InvoiceHTLC,
 
 // updateHtlc aligns the state of an htlc with the given invoice state.
 func updateHtlc(resolveTime time.Time, htlc *InvoiceHTLC,
-	invState ContractState) error {
+	invState ContractState, setID *[32]byte) error {
 
-	switch invState {
+	trySettle := func(persist bool) error {
+		if htlc.State != HtlcStateAccepted {
+			return nil
+		}
 
-	case ContractSettled:
-		if htlc.State == HtlcStateAccepted {
-			htlc.State = HtlcStateSettled
+		// Settle the HTLC if it matches the settled set id. Since we
+		// only allow settling of one HTLC set (for now) we cancel any
+		// that do not match the set id.
+		var htlcState HtlcState
+		if htlc.IsInHTLCSet(setID) {
+			// Non-AMP HTLCs can be settled immediately since we
+			// already know the preimage is valid due to checks at
+			// the invoice level. For AMP HTLCs, verify that the
+			// per-HTLC preimage-hash pair is valid.
+			if setID != nil && !htlc.Preimage.Matches(*htlc.Hash) {
+				return fmt.Errorf("AMP preimage mismatch, "+
+					"preimage=%v hash=%v", htlc.Preimage,
+					htlc.Hash)
+			}
+
+			htlcState = HtlcStateSettled
+		} else {
+			htlcState = HtlcStateCanceled
```

Comment:
It's a bit confusing that a `trySettle` method can take it it `canceled` state...

Reply:
IIUC, this is mean to enforce that if we settle an AMP payment for a given `setID`, then all other accepted HTLCs with a different `setID` are to be cancelled back.

Reply:
Our invoice structure only has one `InvoiceState` field, which means that we can only use this to record the settled state of one HTLCSet. When we move to recurring invoices, we'll need to add an extra layer of indirection for the invoice state so that we can track each set independently. For now we just cancel the HTLC sets that aren't in the set being settled.